README.md
NAnt2
====
Let's chat [![Join the chat at https://gitter.im/NAnt2/NAnt2](https://badges.gitter.im/NAnt2/NAnt2.svg)](https://gitter.im/NAnt2/NAnt2?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)


* [![NAnt2 build](https://github.com/NAnt2/NAnt2/actions/workflows/build.yml/badge.svg)](https://github.com/NAnt2/NAnt2/actions/workflows/build.yml) [![Build status](https://ci.appveyor.com/api/projects/status/jf9itaqnhmjf4aw8/branch/master?svg=true)](https://ci.appveyor.com/project/savornicesei/nant2/branch/master)



[![Maintainability Rating](https://sonarcloud.io/api/project_badges/measure?project=NAnt2_NAnt2&metric=sqale_rating)](https://sonarcloud.io/summary/new_code?id=NAnt2_NAnt2) [![Coverage](https://sonarcloud.io/api/project_badges/measure?project=NAnt2_NAnt2&metric=coverage)](https://sonarcloud.io/summary/new_code?id=NAnt2_NAnt2)
[![CodeScene general](https://codescene.io/images/analyzed-by-codescene-badge.svg)](https://codescene.io/projects/34846)


What is it? 
-----------
NAnt2 is an updated and enhanced fork of [NAnt](https://github.com/nant/nant) - a .NET-based build tool and [NAnt.Contrib](https://github.com/nant/nantcontrib) - NAnt extension. 

It is intended as a drop-in replacement of NAnt, and it is licensed as free and open-source software under the GNU General Public License.

Why NAnt2?
---------
Because people still use it and because it still works great, even if the last release was long time ago (on 9th June 2012).

License
-------
NAnt2 is provided under included GPL license. Portions of it (NAnt2.Contrib) are provided under included LGPL license.  
There is an effort to change license to Apache 2.0 / MIT. If you ever contributed to NAnt, NAnt.Contrib or NAnt2, please state your approval/disapproval on [ticket #14](https://github.com/NAnt2/NAnt2/issues/14)

NAnt and NAnt.Contrib
=====================

What is NAnt?
-------------
In theory it is kind of like `make` without `make`'s wrinkles. In practice it's a lot like Ant.
If you are not familiar with Jakarta Ant you can get more information at the [Ant project web site](http://ant.apache.org/).


Why NAnt?
---------
Because Ant was too Java-specific.

Because Ant needed the Java runtime. NAnt only needs the .NET or Mono runtime.


The Latest Version
------------------
* NAnt - Details of the latest version can be found on the [NAnt sourceforge web site](http://nant.sourceforge.net/)
* Nant.Contrib - Details of the latest version can be found on the [NAnt sourceforge web site](http://nantcontrib.sourceforge.net/)


Files
-----
|  File           |  Purpose                                        |
| --------------- | ----------------------------------------------- |
|  README.md/txt  |  This file                                      |
|  Makefile       |  Makefile for compilation with GNU Make         |
|  Makefile.nmake |  Makefile for compilation with Microsoft NMake  |


Compilation and Installation
-----------------------------
* Windows (net-2.0): [![TeamCity CodeBetter](https://img.shields.io/teamcity/codebetter/bt387.svg)](http://teamcity.codebetter.com/project.html?projectId=NAnt&guest=1)
* Windows (mono-2.0): [![TeamCity CodeBetter](https://img.shields.io/teamcity/codebetter/bt175.svg)](http://teamcity.codebetter.com/project.html?projectId=NAnt&guest=1)
* Linux: [![Travis-ci.org](https://travis-ci.org/nant/nant.svg)](https://travis-ci.org/nant/nant)
Overview
--------
The compilation process uses NAnt to build NAnt.

The approach is to first compile a copy of NAnt (using `make`/`nmake`) for
bootstrapping purpose. Next, the bootstrapped version of NAnt is used in
conjunction with NAnt build file (`NAnt.build`) to build the full version.


Prerequisites
-------------
To build NAnt, you will need the following components:

Windows
-------

* A version of the Microsoft .NET Framework.

  Available from http://msdn.microsoft.com/netframework/
  
  You will need the .NET Framework SDK as well as the runtime 
  components if you intend to compile programs.

  Note: NAnt currently supports versions 1.0, 1.1, 2.0, 3.5, and 4.0 
  of the Microsoft .NET Framework. 

or

* Mono for Windows (version 2.0 or higher)

  Available from http://www.mono-project.com/downloads/

Linux/Unix
----------

* GNU toolchain - including GNU `make`

* `pkg-config`

    Available from: http://www.freedesktop.org/Software/pkgconfig

* A working Mono installation and development libraries (version 2.0 or higher)

    Available from: http://www.mono-project.com/downloads/

        
Building the Software
---------------------
   
Build NAnt using Microsoft .NET:     

* GNU Make

        make install MONO= MCS=csc prefix=<installation path> [DESTDIR=<staging path>] [TARGET=<nant target>]

    eg. `make install MONO= MCS=csc prefix="c:\Program Files" TARGET=net-4.5`

* NMake

        nmake -f Makefile.nmake install prefix=<installation path> [DESTDIR=<staging path>] [TARGET=<nant target>]

    eg. `nmake -f Makefile.nmake install prefix="c:\Program Files" TARGET=net-4.5`


Building NAnt using Mono:

* GNU Make

        make install prefix=<installation path> [DESTDIR=<staging path>] [TARGET=<nant target>]

    eg. `make install prefix="c:\Program Files" TARGET=mono-4.5 MCS="mcs -sdk:4.5"`

* NMake

        nmake -f Makefile.nmake install MONO=mono CSC=mcs prefix=<installation path> [DESTDIR=<staging path>] [TARGET=<nant target>]

    eg. `nmake -f Makefile.nmake install MONO=mono CSC=mcs prefix=/usr/local/`

Note: These instructions only apply to the source distribution of NAnt, as the
binary distribution contains pre-built assemblies.


Documentation
-------------
Documentation is available in HTML format, in the `doc/` directory.


License
========
NAnt is provided under included GPL license.  
NAntContrib is provided under included LGPL license.  
NAnt Copyright (C) 2001-2012 Gerry Shaw  
NAntContrib Copyright (C) 2002 Tomas Restrepo (tomasr@mvps.org)

Contributors
================
NAnt2 (2018 - present day)
---------------------------
* [Simona Avornicesei](https://github.com/savornicesei)

NAnt (2001 - 2018)
------------------
**Lead Developers**
* [Gerry Shaw](https://github.com/gshaw) (gerry_shaw -at- yahoo.com)
* Ian MacLean (ian_maclean -at- another.com)
* Scott Hernandez (scotthernandez -at- hotmail.com)
* [Gert Driesen](https://github.com/drieseng) (gert.driesen -at- ardatis.com)
* [Ryan Boggs](https://github.com/rmboggs)
* [Dominik Guder](https://github.com/dguder)

**Developers**
* Eric V. Smith
* Bernard Vander Beken (bvb -at- iname.com)
* Jason Diamond (jason -at- injektilo.org)
* Jay Turpin (jayturpin -at- hotmail.com)
* Matthew Mastracci (matt -at- aclaro.com)
* John Barstow (John_Barstow -at- gfsg.co.nz )
* Clayton Harbour (claytonharbour -at- sporadicism.com)


**Contributors**
* Aaron A. Anderson (aaron -at- skypoint.com)
* Arjen Poutsma (Arjen -at- ZyLAB.COM)
* Ari Hännikäinen
* Ben Lowery (b3low -at- hotmail.com)
* Bill Baker (bill.baker -at- epigraph.com)
* Brad Wilson (dotnetguy -at- pobox.com)
* Jason Reimer (jason.reimer -at- diversant.net)
* Jeroen Witmond (jnw -at- paralax.nl)
* Joe Jones (joejo -at- microsoft.com)
* Justin Rudd
* Mike Krueger (mike -at- icsharpcode.net)
* Philippe Lavoie
* Sergey Chaban (serge -at- wildwestsoftware.com)
* Shawn Van Ness
* Giuseppe Greco (giuseppe.greco -at- agamura.com)
* John Lam (http://www.iunknown.com/) 

...And more, please see list archives for additional contributors 


NAntContrib (2003 - 2018)
--------------------------
**Lead Developers**
* [Gerry Shaw](https://github.com/gshaw) (gerry_shaw -at- yahoo.com)
* Ian MacLean (ian_maclean -at- another.com)
* Scott Hernandez (scotthernandez -at- hotmail.com)
* [Gert Driesen](https://github.com/drieseng) (gert.driesen -at- ardatis.com)

**Contributors**
* Dominik Guder 
* Clayton Harbour
* Richard Birkby
* Nikhil Gupta 
* Bryce


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## v0.93 - Unreleased

### Added 
- **NAnt.Contrib** is now part of NAnt2 project [NAnt2#77](https://github.com/NAnt2/NAnt2/issues/77)

#### Framework support
- NAnt2 builds now on .NET framework 4.5 that means Windows Vista SP2 and newer and any Linux distribution with Mono v3.0+
- NAnt2 can now be build on Mono 6.12 [NAnt2#68](https://github.com/NAnt2/NAnt2/issues/68), [NAnt2#93](https://github.com/NAnt2/NAnt2/issues/93), [NAnt2#94](https://github.com/NAnt2/NAnt2/issues/94)
- Added support for .NET framework 4.5+ [NAnt2#55](https://github.com/NAnt2/NAnt2/issues/55), [NAnt2#20]([NAnt2#93](https://github.com/NAnt2/NAnt2/issues/20)):
  - net-4.5, 
  - net-4.5.1, 
  - net-4.5.2, 
  - net-4.6, 
  - net-4.6.1, 
  - net-4.6.2, 
  - net-4.7, 
  - net-4.7.1, 
  - net-4.7.2
  - net-4.8
  - mono-4.5
  - mono-4.5.1
- Added support for .NET/Mono Framework 4.5 and VS2012 solution [NAnt2#62](https://github.com/NAnt2/NAnt2/issues/62)
- Updated NAnt to use the newer Windows SDKs by default when targeting net-2.0 before defaulting back to the old .NET 2.0 sdk


#### Core
- Various performance improvements to reduce memory usage, and overall performance during executions of long running tasks [NAnt#PR143](https://github.com/nant/nant/pull/143)
- Upgraded to NUnit 2.6.3
- Properties with null values are no longer allowed
- Implemented parallel execution of targets in the project. [NAnt#PR105](https://github.com/nant/nant/pull/105)
- Fixes builds when make runs multiple jobs simultaneously (`make -j`) [NAnt#PR115](https://github.com/nant/nant/pull/115)

#### Tasks
- Added `assembly::get-informational-version` function
- Added `assembly::get-attribute-property` function to retrieve any custom attribute from the assembly
- Add ability to add password to zip files and to unzip archives protected with passwords
- Added new attribute `expectedexitcode` to `<exec>` task that sets the expected 3rd party tool exit code and fails the build if its not met [NAnt2#63](https://github.com/NAnt2/NAnt2/issues/63)
- Added new function `nant::is-64bit` that checks if NAnt is running in a 64-bit process
- `path::combine` can now combine up to 4 paths [NAnt2#163](https://github.com/NAnt2/NAnt2/issues/163)
- `<trycatch>` task now includes all of the inner exception information in the `catch` property [NAnt#155](https://github.com/nant/nant/issues/155)
- Added new application setting `nant.externalprogram.output.timeout` to control the output/error timeout value for external programs in App.config [NAnt#PR132](https://github.com/nant/nant/pull/132)
- `<echo>Some Important Message</echo>` will not show warning in VS anymore when `nant.xsd` is used to enable build file support [NAnt#PR122](https://github.com/nant/nant/pull/122)
- Added `directory::get-name` function to retrieve directory names from a given path [NAnt#59](https://github.com/nant/nant/issues/59)
- Updated the `csc` task to handle the new compiler model introduced in [Mono 3.0.6](http://www.mono-project.com/Release_Notes_Mono_3.0.6#C.23_Compiler">Mono 3.0) [NAnt#91](https://github.com/nant/nant/issues/91)

### Fixed

#### Core
- Fixed `locatesdk` to read installed SDK information from Wow6432Node as well if NAnt runs as a 64bit process
- Fixed "NET1.1 not found" on x64 systems
- Fixed a null exception in `<locatesdk>` due to absent registry HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Microsoft SDKs\Windows on Windows 10 64bit [NAnt2#97](https://github.com/NAnt2/NAnt2/issues/97) 
- The dollar sign (`$`) can now be escaped by using two dollar signs. (e.g. `<echo message="$${my.prop}"/>` => `"${my.prop}"`) [NAnt#99](https://github.com/nant/nant/issues/99)

#### Tasks
- Fixed streams not being closed in `<mail>` task that could cause errors [NAnt2#66](https://github.com/NAnt2/NAnt2/issues/66)
- Fixed an issue with the `get` task when the request will timeout after the default read write timeout of 5 minutes which may be less than the Timeout value set on the task [NAnt#65](https://github.com/nant/nant/issues/65)
- Fixed code example for `version::get-major-version` function
- Fixed `<copy>` task to also check the `asis` count of a copy file set (next to other "includes") before defaulting to adding all files and subdirectories [NAnt#50](https://github.com/nant/nant/issues/50)
- Fixed a bug in `xmlpoke` task causing additional whitespace and newline character when poke empty value into element [NAnt2#61](https://github.com/NAnt2/NAnt2/issues/61)
- Fixed `<vb6>` task when parsing _ResFile32_ entries in .vbp file [NAnt#PR27](https://github.com/nant/nantcontrib/pull/27)
- Fixed issue where `<vb6>` was looking in the wrong path on x64 systems. [NAnt#PR35](https://github.com/nant/nantcontrib/pull/35)
- Fixed an issue with the `<sysinfo>` task when dealing  with environment property names which contains invalid characters like parentheses. E.g. `ProgramFiles(x86)` on Win7 64 bit will result in `sys.env.ProgramFiles.x86` [NAnt#83](https://github.com/nant/nant/issues/83)
- Fixed an issue with the `<mail>` task when dealing with email address lines ending with a semi-colon [NAnt#61](https://github.com/nant/nant/issues/61)
- `if` and `unless` attributes are evaluated now before trying to expand properties elsewhere within the task [NAnt#110](https://github.com/nant/nant/issues/110)
- Prevent `foreach` from adding properties with null values into the property dictionary [NAnt#84](https://github.com/nant/nant/issues/84)
- Fixed issues related to copying subdirectories that were introduced in NAnt 0.92. ([NAnt#64](https://github.com/nant/nant/issues/64), [NAnt#85](https://github.com/nant/nant/issues/85), and [NAnt#88](https://github.com/nant/nant/issues/88)) 
 

### Deprecated

build_1.ps1
<#
    Main build script for NAnt
#>

#Steps:
#   1. detect OS version; use $IsWindows / !$IsWindows
#   2. detect installed versions of .NET framework, Mono, .NET Core
#   3. create 'bootstrap' and 'bootstrap\lib' folder
#   4. copy all from 'lib' to bootstrap\lib
#   5. copy 'lib\common\neutral\log4net' to 'bootstrap'
#   5. copy 'src\NAnt.Console\App.config' to 'bootstrap\NAnt.exe.config'
#   6. run 'csc  -target:exe -define:NET,NET_1_0,NET_1_1,NET_2_0,NET_3_5,NET_4_0,ONLY_4_0 -out:bootstrap\NAnt.exe -r:bootstrap\log4net.dll  -r:System.Configuration.dll  -recurse:src\NAnt.Console\*.cs src\CommonAssemblyInfo.cs'
#   7. run 'resgen  src\NAnt.Core\Resources\Strings.resx bootstrap\NAnt.Core.Resources.Strings.resources'
#   8. csc  -target:library -warn:0 -define:NET,NET_1_0,NET_1_1,NET_2_0,NET_3_5,NET_4_0,ONLY_4_0 -out:bootstrap\NAnt.Core.dll -r:bootstrap\log4net.dll  -r:System.Web.dll -r:System.Configuration.dll -resource:bootstrap\NAnt.Core.Resources.Strings.resources  -recurse:src\NAnt.Core\*.cs src\CommonAssemblyInfo.cs
#   9. resgen.exe  src\NAnt.DotNet\Resources\Strings.resx bootstrap\NAnt.DotNet.Resources.Strings.resources
#   10. 'csc  -target:library -warn:0 -define:NET,NET_1_0,NET_1_1,NET_2_0,NET_3_5,NET_4_0,ONLY_4_0 -out:bootstrap\NAnt.DotNetTasks.dll  -r:bootstrap\NAnt.Core.dll -r:bootstrap/lib/common/neutral/NDoc.Core.dll  -resource:bootstrap\NAnt.DotNet.Resources.Strings.resources -recurse:src\NAnt.DotNet\*.cs  src\CommonAssemblyInfo.cs'
#   11. 'csc  -target:library -warn:0 -define:NET,NET_1_0,NET_1_1,NET_2_0,NET_3_5,NET_4_0,ONLY_4_0 -out:bootstrap\NAnt.CompressionTasks.dll  -r:bootstrap\NAnt.Core.dll -r:bootstrap\lib\common\neutral\ICSharpCode.SharpZipLib.dll -recurse:src\NAnt.Compression\*.cs src\CommonAssemblyInfo.cs'
#   12. 'csc  -target:library -warn:0 -define:NET,NET_1_0,NET_1_1,NET_2_0,NET_3_5,NET_4_0,ONLY_4_0 -out:bootstrap\NAnt.Win32Tasks.dll  -r:bootstrap\NAnt.Core.dll -r:bootstrap\NAnt.DotNetTasks.dll -r:System.ServiceProcess.dll  -recurse:src\NAnt.Win32\*.cs  src\CommonAssemblyInfo.cs'
#   13. 'bootstrap\NAnt.exe -j  -t:net-4.0 -f:NAnt.build install -D:prefix="d:\Simona Avornicesei\Projects\nant2\build"  -D:destdir="" -D:doc.prefix=""

param(
    [string]$BuildNumber = "0",
    [string]$Version = "0.95"
)

$ErrorActionPreference = 'STOP' #Whenever Write-Error is used, it will produce a terminating error

function Invoke-Build
{
    param(
        [string]$TargetFramework = "mono-4.5"
    )

    switch -Exact ($TargetFramework) 
    {
        "mono-4.0" 
        {  
            make clean build-nant TARGET=mono-4.0 MCS="mcs -sdk:4" build=$BuildNumber version=$Version
            break
        }
        "net-4.0" 
        {  
            make clean build-nant TARGET=net-4.0 MCS="mcs -sdk:4" build=$BuildNumber version=$Version
            break
        }
        "net-4.5" 
        {  
            make clean build-nant TARGET=net-4.5 MCS="mcs -sdk:4.5" build=$BuildNumber version=$Version
            break
        }
        Default 
        {
            make clean build-nant TARGET=mono-4.5 MCS="mcs -sdk:4.5" build=$BuildNumber version=$Version
            break
        }
    }

    if ($LASTEXITCODE -ne 0)
    {
        Write-Error "Error when building nant"
    }
}

function Invoke-Tests
{
    param(
        [string]$TargetFramework = "mono-4.5"
    )

    switch -Exact ($TargetFramework) 
    {
        "mono-4.0" 
        {  
            make clean run-test TARGET=mono-4.0 MCS="mcs -sdk:4" build=$BuildNumber version=$Version
            break
        }
        "net-4.0" 
        {  
            make clean run-test TARGET=net-4.0 MCS="mcs -sdk:4" build=$BuildNumber version=$Version
            break
        }
        "net-4.5" 
        {  
            make clean run-test TARGET=net-4.5 MCS="mcs -sdk:4.5" build=$BuildNumber version=$Version
            break
        }
        Default 
        {
            make clean run-test TARGET=mono-4.5 MCS="mcs -sdk:4.5" build=$BuildNumber version=$Version
            break
        }
    }

    if ($LASTEXITCODE -ne 0)
    {
        Write-Error "Error when running nant tests"
    }
}

Write-Host "[Windows] Build and run tests on mono for version $Version and build number $BuildNumber"

$env:MONO323_HOME = "${env:ProgramFiles(x86)}\Mono-3.2.3"
$env:MONO442_HOME = "${env:ProgramFiles(x86)}\Mono"
$env:MONO64_HOME = "$env:PROGRAMFILES\Mono"

Write-Host "[choco] Installing make"
choco install -yf make
refreshenv
Write-Host "[choco] Uninstall any existing mono packages installed"
choco uninstall mono --all-versions -y -f

Write-Host "--------------------------------"
Write-Host "mono-3.2.2: Building and testing"
Write-Host "--------------------------------"
Write-Host "-- [choco] Installing mono 3.2.3"
choco install mono --version=3.2.3 -yf
Write-Host "-- [env] Add mono to PATH"
$env:PATH += ";$env:MONO323_HOME\bin"
mono --version
Write-Host "-- [make] Build and run tests"
Invoke-Build -TargetFramework net-4.0
Invoke-Tests -TargetFramework net-4.0
Write-Host "-- [choco] Removing mono 3.2.3"
choco uninstall mono --version=3.2.3 -yfn

Write-Host "--------------------------------"
Write-Host "mono-4.4.2: Building and testing"
Write-Host "--------------------------------"
Write-Host "-- [choco] Installing mono 4.4.2"
choco install mono --version=4.4.2 -yf
Write-Host "-- [env] Add mono to PATH"
$env:PATH += ";$env:MONO442_HOME\bin"
mono --version
Write-Host "-- [make] Build and run tests"
Invoke-Build
Invoke-Tests
Write-Host "-- [choco] Removing mono 4.4.2"
choco uninstall mono --version=4.4.2 -yfn

Write-Host "------------------------------------"
Write-Host "mono-5.20.1.34: Building and testing"
Write-Host "------------------------------------"
Write-Host "-- [choco] Installing mono 5.20.1.34"
choco install mono --version=5.20.1.34 -yf
Write-Host "-- [env] Add mono to PATH"
$env:PATH += ";$env:MONO64_HOME\bin"
mono --version
Write-Host "-- [make] Build and run tests"
Invoke-Build
Invoke-Tests

Write-Host "-------------------------------------"
Write-Host "mono-6.12.0.107: Building and testing"
Write-Host "-------------------------------------"
Write-Host "-- [choco] Installing mono 6.12.0.107"
choco upgrade mono --version=6.12.0.107 -yf
mono --version
Write-Host "-- [make] Build and run tests"
Invoke-Build
Invoke-Tests